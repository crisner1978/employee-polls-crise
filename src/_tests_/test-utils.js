import React from 'react'
import { render } from '@testing-library/react'
import { configureStore } from '@reduxjs/toolkit'
import { Provider } from 'react-redux'
import authReducer from '../features/authSlice'
import questionsReducer from '../features/questionsSlice'
import usersReducer from '../features/usersSlice'

export function renderWithProviders(
  ui,
  {
    preloadedState = {},
    store = configureStore({
      reducer: { auth: authReducer, users: usersReducer, questions: questionsReducer },
      preloadedState,
    }),
    ...renderOptions
  } = {}
) {
  function Wrapper({ children }) {
    return <Provider store={store}>{children}</Provider>
  }
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) }
}
