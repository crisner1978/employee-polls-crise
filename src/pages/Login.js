import React from "react";
import { FcGoogle } from "react-icons/fc";
import { FaGithub } from "react-icons/fa";
import db, { auth, googleProvider, githubProvider } from "../lib/firebase";
import { signInWithPopup } from "firebase/auth";
import { collection, getDocs, query, where } from "firebase/firestore/lite";
import { useNavigate } from "react-router-dom";

export default function Login({ setNewUser, setUser }) {
  const navigate = useNavigate();

  async function signIn(provider) {
    const data = await signInWithPopup(auth, provider);
    if (data) {
      await checkUsername(data.user.uid);
      setUser(data.user);
      navigate("/", { replace: true });
    }
  }

  async function checkUsername(uid) {
    const col = collection(db, "usernames");
    const q = query(col, where("uid", "==", uid));
    const { empty } = await getDocs(q);
    setNewUser(empty);
  }

  return (
    <div className="w-80 mx-auto">
      <header className="pt-20 mb-8 text-center">
        <h1 className="font-bold text-xl">Sign Up for Employee Polls</h1>
      </header>

      <div className="space-y-4">
        <LoginOption
          Icon={FcGoogle}
          onClick={() => signIn(googleProvider)}
          text="Continue with Google"
        />
        <LoginOption
          Icon={FaGithub}
          onClick={() => signIn(githubProvider)}
          text="Continue with Github"
        />
      </div>
    </div>
  );
}

function LoginOption({ Icon, onClick, text }) {
  return (
    <div
      className="flex items-center cursor-pointer bg-white hover:bg-gray-100 rounded-[5px] border border-gray-300 h-14"
      onClick={onClick}>
      <div className="rounded-l-[5px] px-3">
        <Icon className="h-5 w-5 mx-2" />
      </div>
      <div className="border-l border-gray-300 rounded-r-[5px] p-4 h-12 w-full flex justify-center items-center">
        <h3 className="font-medium">{text}</h3>
      </div>
    </div>
  );
}
