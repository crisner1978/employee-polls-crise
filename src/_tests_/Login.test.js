import { screen } from '@testing-library/react'
import * as React from 'react'
import { MemoryRouter } from 'react-router-dom'
import Login from '../pages/Login'
import { renderWithProviders } from './test-utils'

describe('Login', () => {
  it('will match snapshot', () => {
    var view = renderWithProviders(
      <MemoryRouter>
        <Login />
      </MemoryRouter>
    )
    expect(view).toMatchSnapshot()
  })
})

describe('Login Events', () => {
  it('will display an error message if username is less than 5 characters', async () => {
    renderWithProviders(
      <MemoryRouter>
        <Login />
      </MemoryRouter>,
      {
        preloadedState: {
          users: {
            sarahedo: {
              id: 'sarahedo',
              password: 'password123',
              name: 'Sarah Edo',
              avatarURL: null,
              answers: {
                '8xf0y6ziyjabvozdd253nd': 'optionOne',
                '6ni6ok3ym7mf1p33lnez': 'optionOne',
                am8ehyc8byjqgar0jgpub9: 'optionTwo',
                loxhs1bqm25b708cmbf3g: 'optionTwo',
                xj352vofupe1dqz9emx13r: 'optionOne',
              },
              questions: ['8xf0y6ziyjabvozdd253nd', 'am8ehyc8byjqgar0jgpub9'],
            },
          },
        },
      }
    )

    var title = screen.getByTestId('login_loading')
    expect(title).toBeInTheDocument()
    expect(title).toHaveTextContent('Loading...')
  })
})
