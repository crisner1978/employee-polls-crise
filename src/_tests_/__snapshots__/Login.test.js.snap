// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login will match snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="mx-auto w-80"
      >
        <header
          class="mb-4 pt-20 text-center"
        >
          <h1
            class="border-b-4 pb-2 text-2xl font-bold"
            data-testid="login_page"
          >
            Employee Polls Login Portal
          </h1>
        </header>
        <h2
          class="mb-4 border-b-2 pb-2 text-center text-xl font-medium shadow-sm shadow-black/30"
        >
          Please select a user to login
        </h2>
        <div
          class="space-y-4"
        />
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="mx-auto w-80"
    >
      <header
        class="mb-4 pt-20 text-center"
      >
        <h1
          class="border-b-4 pb-2 text-2xl font-bold"
          data-testid="login_page"
        >
          Employee Polls Login Portal
        </h1>
      </header>
      <h2
        class="mb-4 border-b-2 pb-2 text-center text-xl font-medium shadow-sm shadow-black/30"
      >
        Please select a user to login
      </h2>
      <div
        class="space-y-4"
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "store": Object {
    "@@observable": [Function],
    "dispatch": [Function],
    "getState": [Function],
    "replaceReducer": [Function],
    "subscribe": [Function],
  },
  "unmount": [Function],
}
`;
